go/batch-submitter/drivers/proposer/driver.go

```diff
@@ -121,24 +126,35 @@ func (d *Driver) SubmitBatchTx(
 	ctx context.Context,
 	start, end, nonce, gasPrice *big.Int) (*types.Transaction, error) {
 
-	var blocks []*l2types.Block
+	name := d.cfg.Name
+
+	batchTxBuildStart := time.Now()
+
+	var (
+		stateRoots         [][32]byte
+		totalStateRootSize uint64
+	)
 	for i := new(big.Int).Set(start); i.Cmp(end) < 0; i.Add(i, bigOne) {
 		block, err := d.cfg.L2Client.BlockByNumber(ctx, i)
 		if err != nil {
 			return nil, err
 		}
 
-		blocks = append(blocks, block)
-
-		// TODO(conner): remove when moving to multiple blocks
-		break //nolint
-	}
+		// Consume state roots until reach our maximum tx size.
+		if totalStateRootSize+stateRootSize > d.cfg.MaxTxSize {
+			break
+		}
+		totalStateRootSize += stateRootSize
 
-	var stateRoots = make([][32]byte, 0, len(blocks))
-	for _, block := range blocks {
 		stateRoots = append(stateRoots, block.Root())
 	}
 
+	batchTxBuildTime := float64(time.Since(batchTxBuildStart) / time.Millisecond)
+	d.metrics.BatchTxBuildTime.Set(batchTxBuildTime)
+	d.metrics.NumTxPerBatch.Observe(float64(len(stateRoots)))
```

Comment:
Perhaps the metric would be better named as `NumElementsPerBatch` but its ok as is